proposals/enum-entries.md

```diff
@@ -0,0 +1,174 @@
+To have an ability to further extend Enum's API in a non-intrusive manner that does not involve code-generation
+for each enum, we expose a direct subtype of `List<E>` named `EnumEntries`.
+
+All potential extensions, such as `valueOfOrNull(String)`, `hasValue(String)` can be implemented on the standard library
+level as extensions of `EnumEntries`.
+
+### Naming alternatives
+
+Various languages use different notations to refer to enumeration elements, making the naming choice especially hard. 
+
+We have considered the following alternatives:
+
+* `values` property. While being the most familiar one, it carries the burdens of the existing `values()` API:
+  * It is easy to misspel it with `values()` and fall into the same trap.
+  * Introduction of `values` property adds an unfixable method reference ambiguity for all the callers of `E::values`.
+* Java language [refers](https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.9.1) to enum elements as `constants`. 
+  * The risks of introducing widely-used API named `constants` with the potential clashes with [Kotlin constant evaluation](https://youtrack.jetbrains.com/issue/KT-14652) outweighs the benefits of the straightforward name.
+* `valuesList` and `entryList` suffer the same disease -- excessive verbosity of the name.
+* `components` is already reserved in Kotlin for positional deconstructors.
+* Kotlin specification and documentation already refer to enum elements as entries, making it our final choice.
+
+### Translation strategy for Kotlin enums
+
+For source-compiled enums, a new static property with a JVM signature `public static EnumEntries<E> getEntries()` is added,
+returning a pre-allocated immutable list of enum entries.
+
+For the sake of compatibility with bytecode pre-processors and reflection usages that modify a backing array `$VALUES` field,
+the generated bytecode for `values()` is left intact, and the resulting list of entries reads the underlying `$VALUES` lazily,
+leaving the chance for existing bytecode preprocessors to change the array without dealing with list allocation.
+
+#### Implementation note
+
+A special type, `EnumEntriesList` is introduced to the standard library. The type has the only constructor
+accepting the functional type with a signature `() -> Array<E>`. The corresponding lambda is generated
+using the standard `invokedynamic` with `LambdaMetafactory`.
+
+The final decompiled class for the enum
+```kotlin
+enum class MyEnum {
+    A;
+}
+```
+
+has the following form (all irrelevant parts omitted):
+```java
+enum MyEnum extends Enum<MyEnum> {
+    private static final synthetic MyEnum[] $VALUES
+    private static final synthetic EnumEntries<MyEnum> $ENTRIES;
+   
+    <clinit> {
+        A = new MyEnum("A", 0);
+        $VALUES = $values();
+        Supplier<MyEnum[]> supplier = #invokedynamic ..args.. values;
+        $ENTRIES = new EnumEntriesList(supplier);
```

Comment:
Now it's a call to the `enumEntries` function instead of the constructor of `EnumEntries`